osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java

```diff
@@ -1,20 +1,196 @@
+                   final Date previousLogTime,
+                   final ProfileCaptureTime intervalTime,
+                   final DlmsHelper dlmsHelper)
+            throws ProtocolAdapterException, BufferedDateTimeValidationException {
+
+        final PeriodTypeDto queryPeriodType = periodicMeterReadsQuery.getPeriodType();
+        final DateTime from = new DateTime(periodicMeterReadsQuery.getBeginDate());
+        final DateTime to = new DateTime(periodicMeterReadsQuery.getEndDate());
+
+        final Integer clockIndex = attributeAddressForProfile.getIndex(DlmsObjectType.CLOCK, null);
+
+        CosemDateTimeDto cosemDateTime = null;
+
+        if (clockIndex != null) {
+            cosemDateTime = dlmsHelper.readDateTime(bufferedObjects.get(clockIndex),
+                    "Clock from " + queryPeriodType + " buffer");
+        }
+
+        final DateTime bufferedDateTime = cosemDateTime == null ? null : cosemDateTime.asDateTime();
+
+        if (bufferedDateTime != null) {
+            dlmsHelper.validateBufferedDateTime(bufferedDateTime, cosemDateTime, from, to);
+
+            getLogger().info("Calculated date based on COSEM date {}", bufferedDateTime.toDate());
+            getLogger().info("Timezoneless date == " + LocalDateTime.ofInstant(bufferedDateTime.toDate().toInstant(), ZoneOffset.systemDefault()));
+            return bufferedDateTime.toDate();
+        } else {
+            // no date was available, calculate date based on previous value
+            return calculateIntervalDate(periodicMeterReadsQuery.getPeriodType(), previousLogTime, intervalTime);
+        }
+    }
+
+    /**
+     * Calculates/derives the next interval date in case it was not present in the current meter read record.
+     *
+     * @param periodTypeDto   the time interval period.
+     * @param previousLogTime the logTime of the previous meter read record
+     * @param intervalTime    the interval time for this device to be taken into account when the periodTypeDto is INTERVAL
+     * @return the derived date based on the previous meter read record, or null if it cannot be determined
+     */
+    private Date calculateIntervalDate(final PeriodTypeDto periodTypeDto,
+                                       final Date previousLogTime,
+                                       final ProfileCaptureTime intervalTime) throws BufferedDateTimeValidationException {
+
+        if (previousLogTime == null) {
+            return null;
+        }
+
+        switch (periodTypeDto) {
+            case DAILY:
+                return Date.from(previousLogTime.toInstant().plus(Duration.ofDays(1)));
+            case MONTHLY:
+                LocalDateTime localDateTime = LocalDateTime.ofInstant(previousLogTime.toInstant(), ZoneId.systemDefault()).plusMonths(1);
+
+                return Date.from(localDateTime.atZone(ZoneId.systemDefault())
+                        .toInstant());
+            case INTERVAL:
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce this switch case number of lines from 7 to at most 5, for example by extracting code into methods. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1151)

Reply:
done